### DISCUSSION TITLE
Is it possible to use a node library on my application?

### DISCUSSION THREAD

**Discussion by fernandotetu** [Moddable-OpenSource/moddable#911] [General]
I would like to ask about the possibility of using a node library on my project. I'm assuming that my hardware there are capabilities to support it.

In particular, I would like to use [cron library](https://www.npmjs.com/package/cron). My intention is to execute an action every day at a specific time. For that, I have downloaded the library and his dependency and point for the node_module folder on my manifest file (I try diferent entry and paths). However, I always received a message: "import CronJob not found!"

I would really appreciate it if anybody could help me to use node libraries with Moddable

**Comment by wilberforce** [comment]
post your manifest file and your main.js

**Reply by fernandotetu** [reply]
Hi @wilberforce , thanks for your assistance. 

The main file:

	import { CronJob } from 'cron';
	trace( 'Success' + "\n");

The manifest file:

	{
		"include": "$(MODDABLE)/examples/manifest_base.json",
		"modules": {
			"*": [
				"./main"
			],
			"cron":"$(MODDABLE)/node_modules/cron/lib/*"
		}
	}

I put on the Moddable's  root directory a folder with the npm library; this is its structure:
![image](https://user-images.githubusercontent.com/3038038/170495903-184d2d9b-a93c-451f-a077-08ab1f9cfa2b.png)

I have been tried different entries to alias "cron". At the file cron.js there is the CronJob export.

**Comment by phoddie** [comment]
As noted in the New Issue template:

> Questions opened as an issue will be converted to a Discussion.

Moving to Discussion.

**Comment by phoddie** [comment]
Node.js is, more or less, a JavaScript runtime for POSIX. The Moddable SDK is also a JavaScript runtime, however it not built on POSIX because it targets microcontrollers which generally use a lightweight RTOS. Because the XS JavaScript engine in the Moddable SDK implements the same standard programming language (recent edition of ECMA-262) as the V8 engine used in Node.js, the Moddable SDK can run many standard ECMAScript Modules (ESM) that are written for Node.js. Many developers using the Moddable SDK have done this successfully.

However, when a Node package depends on APIs or behaviors unique to Node.js runtime, they generally do not work. While Node.js and the Moddable SDK share a programming language, they do not share APIs or behaviors beyond that.

The Node.js  [cron package](https://www.npmjs.com/package/cron) you mention appears to depend on both the Node.js `child_process` [API](https://nodejs.org/api/child_process.html) and the [Luxon](https://moment.github.io/luxon/#/) package (which in turn seems to depend on Intl?). That looks like a problem. It may be possible to adapt code to eliminate or replace these dependencies. However, you'll have a large body of code to solve a reasonably small problem. You can probably get the result you need with a few calculations and the Moddable SDK `Timer` [API](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/base/base.md#timer).

**Reply by fernandotetu** [reply]
Thank you @phoddie 

Now it's clear. This library is not compatible. But, assuming I have found a library compatible, what are the steps to integrate with the Moddable application?

Regarding my concern, I am writing a paper to prove that JavaScript can be better for coding embedded software than the C language. For example, using the Monitor class for monitoring a digital pin is 35% more energy-efficient than the polling strategy. 
My intention with the scheduled task was to demonstrate how simple it could be to code it compared to the C language. For this reason, I chose to use the third library.

**Comment by phoddie** [comment]
> For example, using the Monitor class for monitoring a digital pin is 35% more energy-efficient than the polling strategy.

That's a nice number. I assume that's an Espressif part. On silicon with better sleep support, I think the difference is even more significant.

> My intention with the scheduled task was to demonstrate how simple it could be to code it compared to the C language. For this reason, I chose to use the third library.

I understand. In that case, it could be acceptable to use a different library instead of the cron package? If so, it would be straightforward to provide some of the same functionality with about the same API but a different implementation. That would be useful to have for embedded developers (I hadn't seen cron before!) What do you think?

**Reply by fernandotetu** [reply]
Hey there! 

>	I understand. In that case, it could be acceptable to use a different library instead of the cron package? If so, it would be straightforward to provide some of the same functionality with about the same API but a different implementation. That would be useful to have for embedded developers (I hadn't seen cron before!) What do you think?

For sure, it'll be useful. I'll work to implement a time-based algorithm to provide cron functionalities. So when I get some results, I come back here to share!

**Reply by AlexLucardGoodman** [reply]
Hi there! Im sorf of in the same place as you were. Im trying to get a cron like fuctionallity to work with the Moddable SDK. Any luck with that? any recommended libraries that would be compatible with the moddable sdk? thank you

**Reply by phoddie** [reply]
Given the dependencies of the Node.js cron library, it seems unlikely that it will be a trivial port. It should be straightforward to put a subset of that in place. I actually need that for [Node-RED MCU Edition](https://github.com/phoddie/node-red-mcu) because the Node-RED Inject node has features explicitly based on cron, *"The "Interval between times" and "at a specific time" options use the standard cron system..."*.  Here's one of its configuration dialogs:

<img width="508" alt="image" src="https://user-images.githubusercontent.com/6856458/235247547-cf63b9c0-b36e-49da-b244-14f547a6402d.png">

That defines my feature requirements. ;) What precisely do you need?

**Reply by phoddie** [reply]
I had a little time today, so I went back and looked at at Cron. I remember it used Moment but it seems to have been updated to use Luxon. That looked a little less complicated. With a little hacking, I was able to get Cron and Luxon running with the Moddable SDK! In my quick tests it is working nicely.

My Moddable SDK fork of Cron and Luxon is [here](https://github.com/phoddie/node-cron#cron-for-moddable-sdk). it includes notes on:

- How to use the port in your projects
- Changes made to the modules
- Likely limitations

**Reply by AlexLucardGoodman** [reply]
you're a genius! just tested it and works at a first glance. I will be testing it and get back on any weird stuff i find... it would be awesome to have a native cron builtin the Moddable sdk. Im reading your notes and i understand moddable uses ESM vs CommonJs and that needs to be adjusted in order to use a third party library, but Im a bit confused about the port thing you mention... is it a real port or youre talking about portability? thank you!

**Reply by phoddie** [reply]
> just tested it and works at a first glance. I will be testing it

Cool! That would be great.

>  it would be awesome to have a native cron builtin the Moddable sdk

I don't know that "builtin" is essential. Including the module is easy enough. But, Cron + Luxor is bigger than necessary, mostly because Luxor provides lots of feature unused by Cron. Shrinking that would be nice.

A naive implementation of a subset Cron is easy enough and would be smaller. But it would likely also be frustrating to use because of limitations. Using the battle-tested Cron Node module seems practical for now. If someone wants to create a streamlined implementation, they surely can -- and if they implement the same API, it will just slide in.

> Im reading your notes... Im a bit confused about the port thing you mention... is it a real port or youre talking about portability?

The [Cron Port](https://github.com/phoddie/node-cron#cron-port) and [Luxon Port](https://github.com/phoddie/node-cron#luxon-port) sections of the document summarize changes made to the source code of those modules so to be compatible with the Moddable SDK runtime. I included those notes to perhaps help others bringing module from Node to Embedded JavaScript.

**Reply by jermsam** [reply]
I understand that the issue was resolved for this specific need, but there are plenty of cool nodejs libraries that were built  with commonjs some of which have very many peer dependencies written with the same standard of js... what's the way forward? What are the steps to using a node library written with commonjs?

**Reply by jermsam** [reply]
There is a package called to-esm that converts commonjs modules to mjs modules, hopefully by adding support for .mjs, this entire non compatibility issue will be solved

**Reply by phoddie** [reply]
Please see note about `.mjs` support [here](https://github.com/Moddable-OpenSource/moddable/discussions/1066#discussioncomment-7417855).


That said, CommonJS is not supported directly so you will need to convert them to ESM, either manually or with some tool. Once that is done, packages that use only standard JavaScript will generally work. But, packages that depend on Node builtins generally will not as those dependencies are not present on microcontrollers.

**Reply by jermsam** [reply]
@phoddie what about using https://mongoose-os.com ?
